libazureinit/src/goalstate.rs

```diff
@@ -41,34 +45,69 @@ pub struct RoleInstance {
     instance_id: String,
 }
 
-pub async fn get_goalstate(client: &Client) -> Result<Goalstate, Error> {
-    let url = "http://168.63.129.16/machine/?comp=goalstate";
+static DEFAULT_GOALSTATE_URL: &str =
+    "http://168.63.129.16/machine/?comp=goalstate";
+
+pub async fn get_goalstate(
+    client: &Client,
+    retry_interval: Duration,
+    total_timeout: Duration,
+    input_url: Option<&str>,
+) -> Result<Goalstate, Error> {
+    let url = match input_url {
+        Some(url) => url,
+        None => DEFAULT_GOALSTATE_URL,
+    };
 
     let mut headers = HeaderMap::new();
     headers.insert("x-ms-agent-name", HeaderValue::from_static("azure-init"));
     headers.insert("x-ms-version", HeaderValue::from_static("2012-11-30"));
 
-    let request = client.get(url).headers(headers);
-    let response = request.send().await?;
+    let response = timeout(total_timeout, async {
+        loop {
+            if let Ok(response) = client
+                .get(url)
+                .headers(headers.clone())
+                .timeout(Duration::from_secs(30))
+                .send()
+                .await
+            {
+                let statuscode = response.status();
 
-    if response.status().is_success() {
-        let body = response.text().await?;
+                if statuscode.is_success() && statuscode == StatusCode::OK {
+                    return response.error_for_status();
```

Comment:
Since there's a check on the status code before we get here this is equivalent to `return Ok(response)`, which I think is a little clearer since it doesn't require the reader to know `error_for_status()` returns `Result<Self>` and can only ever return `Ok` in this case.